c10/test/util/complex_test_common.h

```diff
@@ -0,0 +1,461 @@
+  constexpr scalar_t zero = scalar_t();
+  static_assert(c10::complex<scalar_t>(num1, num2).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(num1, num2).imag() == num2, "");
+  static_assert(c10::complex<scalar_t>(num1).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(num1).imag() == zero, "");
+  static_assert(c10::complex<scalar_t>().real() == zero, "");
+  static_assert(c10::complex<scalar_t>().imag() == zero, "");
+}
+
+template<typename scalar_t, typename other_t>
+C10_HOST_DEVICE void test_construct_from_other() {
+  constexpr other_t num1 = other_t(1.23);
+  constexpr other_t num2 = other_t(4.56);
+  constexpr scalar_t num3 = scalar_t(num1);
+  constexpr scalar_t num4 = scalar_t(num2);
+  static_assert(c10::complex<scalar_t>(c10::complex<other_t>(num1, num2)).real() == num3, "");
+  static_assert(c10::complex<scalar_t>(c10::complex<other_t>(num1, num2)).imag() == num4, "");
+}
+
+MAYBE_GLOBAL void test_convert_constructors() {
+  test_construct_from_scalar<float>();
+  test_construct_from_scalar<double>();
+
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(!std::is_convertible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_convertible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  test_construct_from_other<float, float>();
+  test_construct_from_other<float, double>();
+  test_construct_from_other<double, float>();
+  test_construct_from_other<double, double>();
+}
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_construct_from_std() {
+  constexpr scalar_t num1 = scalar_t(1.23);
+  constexpr scalar_t num2 = scalar_t(4.56);
+  static_assert(c10::complex<scalar_t>(std::complex<scalar_t>(num1, num2)).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(std::complex<scalar_t>(num1, num2)).imag() == num2, "");
+}
+
+MAYBE_GLOBAL void test_std_conversion() {
+  test_construct_from_std<float>();
+  test_construct_from_std<double>();
+}
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template<typename scalar_t>
+void test_construct_from_thrust() {
+  constexpr scalar_t num1 = scalar_t(1.23);
+  constexpr scalar_t num2 = scalar_t(4.56);
+  ASSERT_EQ(c10::complex<scalar_t>(thrust::complex<scalar_t>(num1, num2)).real(), num1);
+  ASSERT_EQ(c10::complex<scalar_t>(thrust::complex<scalar_t>(num1, num2)).imag(), num2);
+}
+#endif
+
+TEST(TestConstructors, FromThrust) {
+#if defined(__CUDACC__) || defined(__HIPCC__)
```

Comment:
should we add this check outside the test:
```#if defined(__CUDACC__) || defined(__HIPCC__)
TEST(..) {
...
}
#endif``` so that we don't have to add this check separately for any other tests that we add in future

Reply:
fixed